### DISCUSSION TITLE
Clarification on trace_paths() output when a path has reflections AND diffractions

### DISCUSSION THREAD

**Discussion by jmoraispk** [NVlabs/sionna#276] [General]
Dear Sionna Team,

Thanks for another great release!

A quick question on the output of trace_paths(). It outputs the specular reflected, the scattered, and the diffracted paths _separately_, along with the tmp variables of each. Suppose a single path between tx and rx has "different types of interaction with the primitives," for example, reflections and diffractions. How is the output of this function structured? Is this path included in both outputs? Or are the paths segmented in some way? 

Hopefully, you can help clarify, I couldn't quite determine this from the source.

**Comment by faycalaa** [comment]
Hi,

The ray tracer of Sionna currently traces four types of paths for computing channel impulse responses:

- Line of sight
- Specular chains, that consist only of specular reflections, labelled as the "specular paths"
- Scattered paths, that consist of a prefix of `k` specular reflections (`0 <= k <= max_depth-1`) followed by a diffuse reflection
- Diffracted paths, that consist of a single diffraction (first-order diffraction)

The six outputs of the `trace_paths()` function correspond to these three types of paths, plus the tmp variables of each. The ray tracer currently does not trace paths that are reflected and diffracted.

Note that this is slightly different for coverage map computation, for which path tracing and EM field computation are not separated into different functions, and which traces the following types of paths:

- Line of sight
- Reflected paths, that consist of specular and/or diffuse reflections, in any order
- Diffracted paths, that consist of a single diffraction (first-order diffraction)

Hope this helps.

**Reply by jmoraispk** [reply]
Thank you. Yes, it helps! But it also raises another question: What obstacles prevent Sionna from supporting all possible paths? Is it a ray tracing solver implementation problem? A complexity problem? A pragmatic approach since these are the most important anyway? (Are they?)

**Reply by jhoydis** [reply]
We always need to strike a balance between performance and accuracy. Adding higher-order diffraction (or specular reflections + diffraction) would increase the ray tracing complexity significantly without providing substantial improvements in accuracy for most scenarios that we are currently interested in. For this reason, Sionna does not compute all possible paths.

**Reply by Liu-Eroteme** [reply]
Hey there! I'm currently trying to help a colleague out and find a suitable alternative for the ray tracing simulations he's doing using altair winprop, and so far sionna seems promising! However, from what i know, second order diffraction is something we care about for one reason or another, so i wanted to ask how much of the ground work would already be there?
